services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -0,0 +1,241 @@
+//            }
+
+            // Verify/Allocate Gateway IP, subnet id, port id, subnet cidr, response:IP - unique
+//            IPStateJson ipResponse = this.subnetService.allocateIPGateway(inSubnetState.getId(), inSubnetState.getCidr(), portId);
+//            if (ipResponse == null) {
+//                throw new ResourcePersistenceException();
+//            }
+
+            // set up value of properties for subnetState
+            List<RouteWebObject> routes = subnetState.getRoutes();
+            if (routes == null) {
+                routes = new ArrayList<>();
+            }
+            routes.add(routeResponse.getRoute());
+            subnetState.setRoutes(routes);
+            //subnetState.setGatewayIp(ipResponse.getIpState().getIp());
+
+        } catch (ResourcePersistenceException e) {
+            throw new Exception(e);
+        } catch (FallbackException e) {
+            // Route info of subnet rollback
+            List<RouteWebObject> routes = resource.getSubnet().getRoutes();
+            if (routes != null) {
+                RouteWebObject route = routes.get(routes.size() - 1);
+                this.subnetService.routeRollback(route.getId(), resource.getSubnet().getVpcId());
+            }
+            throw new Exception(e);
+        }
+
+        return new SubnetStateJson(subnetState);
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/vpcs/{vpcid}/subnets/{subnetid}", "v4/{projectid}/vpcs/{vpcid}/subnets/{subnetid}"})
+    public SubnetStateJson updateSubnetState(@PathVariable String projectid, @PathVariable String vpcid, @PathVariable String subnetid, @RequestBody SubnetStateJson resource) throws Exception {
+
+        SubnetState subnetState = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(vpcid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(subnetid);
+            SubnetState inSubnetState = resource.getSubnet();
+            RestPreconditionsUtil.verifyResourceNotNull(inSubnetState);
+            RestPreconditionsUtil.populateResourceProjectId(inSubnetState, projectid);
+            RestPreconditionsUtil.populateResourceVpcId(inSubnetState, vpcid);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(subnetid);
+            if (subnetState == null) {
+                throw new ResourceNotFoundException("Subnet not found : " + subnetid);
+            }
+
+            RestPreconditionsUtil.verifyParameterEqual(subnetState.getProjectId(), projectid);
+            RestPreconditionsUtil.verifyParameterEqual(subnetState.getVpcId(), vpcid);
+
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+            subnetState = this.subnetDatabaseService.getBySubnetId(subnetid);
+
+        } catch (ParameterNullOrEmptyException e) {
+            throw new Exception(e);
```

Comment:
You might want to log different message in different logging level here.

Reply:
NP